### PR TITLE
UG-812: Implement create list and save article overlay variant

### PR DESCRIPTION
**Pull request by meel-io**
### Description
 This PRs implements the "Create list and save article overlay" variant that will be shown on [this US Growth test](https://www.figma.com/file/nNlBgqIg0aXAOrYLrxPmSg/Professors-Lists?node-id=500%3A28355)

Its main changes consist of:

- Adding a data property  `data-myft-ui-variant=createListAndSaveArticleVariant` to the save for later button that will determine if the test is active based on the value of the professorLists flag.
- Adding a check for that data property in the `myft.user.saved.content.add` event listener, and displays the new overlay variant if the user has no lists.
- Implement the overlay variant using `o-overlay` and implement methods to: position the overlay next to the `SaveForLater` component, reposition the overlay on resize or scroll (mobile only) handle the interactions to create a new list and add/remove the article to lists.

### Screenshots

Screenshot 2022-04-13 at 10 31 00
Screenshot 2022-04-13 at 10 31 20

### How to test this code?

- Check out this branch locally and run: 
  - `npm link`
  - `bower link`
- Check out `next-article` and run:
  - `npm link @financial-times/n-myft-ui`
  - `bower link n-myft-ui`
- Build and run `next-article`
  - `make build run`
- Go to FT Toggler and select the variant on the [professorLists flag](https://toggler.ft.com/#professorLists)
- Go to [myFT section](https://www.ft.com/myft/lists) and make sure you don't have any lists created. 
- Go to [an article](https://local.ft.com:5050/content/aa9caaad-c19f-4451-b13b-e514b9c284a5) and click on the `Save`  button on the Share Navigation bar
- Confirm that you can: create a new list, add the article to the list and remove the article from the list. These changes must be reflected on your [myFT lists page](https://www.ft.com/myft/lists/)
- Test it on the following article types
  - [x] Normal articles ([Local](https://local.ft.com:5050/content/395650fa-5b9c-11e5-a28b-50226830d644) | [Live](https://www.ft.com/content/395650fa-5b9c-11e5-a28b-50226830d644) | `395650fa-5b9c-11e5-a28b-50226830d644`)
  - [x] Articles with toppers ([Local](https://local.ft.com:5050/content/fe291f16-4aa6-11e7-a3f4-c742b9791d43) | [Live](https://www.ft.com/content/fe291f16-4aa6-11e7-a3f4-c742b9791d43) | `fe291f16-4aa6-11e7-a3f4-c742b9791d43`)
  - [x] Package Intro page
([Local](https://local.ft.com:5050/content/de4d756c-04be-11e7-ace0-1ce02ef0def9) | [Live](https://www.ft.com/content/de4d756c-04be-11e7-ace0-1ce02ef0def9) | `de4d756c-04be-11e7-ace0-1ce02ef0def9`)
  - [x] Package article page ([Local](https://local.ft.com:5050/content/bb1d8246-75e1-11e7-a3e8-60495fe6ca71) | [Live](https://www.ft.com/content/bb1d8246-75e1-11e7-a3e8-60495fe6ca71) | `bb1d8246-75e1-11e7-a3e8-60495fe6ca71`)
  - [x] Kitchen Sink Components ([Local](https://local.ft.com:5050/content/146da558-4dee-11e3-8fa5-00144feabdc0) | [Live](https://www.ft.com/content/146da558-4dee-11e3-8fa5-00144feabdc0) | `146da558-4dee-11e3-8fa5-00144feabdc0`)

### CODE REVIEW

File: myft/ui/save-article-to-list-variant.js [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+function stringToHTMLElement (string) {
+	const template = document.createElement('template');
+	template.innerHTML = string.trim();
+	return template.content.firstChild;
+}
+
+function FormElement (createList) {
+	const formString = `
+	<form class='myft-ui-create-list-variant-form'>
+		<label class="o-forms-field">
```
Comment (jamesnicholls) [inline myft/ui/save-article-to-list-variant.js:143]: **question:** Doesn't the `label` element need some text?

File: myft/ui/save-article-to-list-variant.js [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+function ListsElement (lists, addToList, removeFromList) {
+	const currentList = document.querySelector('.myft-ui-create-list-variant-lists');
+	if (currentList) {
+		currentList.remove();
+	}
+
+	const listCheckboxElement = ListCheckboxElement(addToList, removeFromList);
+
+	const listsTemplate = `
+	<div class="myft-ui-create-list-variant-lists o-forms-field o-forms-field--optional" role="group" aria-labelledby="checkbox-group-title" aria-describedby="checkbox-group-info">
```
Comment (jamesnicholls) [inline myft/ui/save-article-to-list-variant.js:196]: **question:** Do we need `aria-labelledby` and `aria-describedby` on this element? The IDs referenced in these attributes don't exist on the page.

File: myft/ui/save-article-to-list-variant.js [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+			${!lists.length ? '<p class="myft-ui-create-list-variant-add-description">Lists are a simple way to curate your content</p>' : ''}
+		</div>
+	`;
+
+	return stringToHTMLElement(content);
+}
+
+function HeadingElement () {
+	const heading = `
+		<span class="myft-ui-create-list-variant-heading">Added to <a href="https://www.ft.com/myft/saved-articles">saved articles</a> in <span class="myft-ui-create-list-variant-icon"><span class="myft-ui-create-list-variant-icon-visually-hidden">my ft</span></span></span>
```
Comment (gyss) [inline myft/ui/save-article-to-list-variant.js:181]:
```suggestion
		<span class="myft-ui-create-list-variant-heading">Added to <a href="https://www.ft.com/myft/saved-articles">saved articles</a> in <span class="myft-ui-create-list-variant-icon"><span class="myft-ui-create-list-variant-icon-visually-hidden">my FT</span></span></span>
```

This allows VoiceOver to read the characters correctly

File: myft/ui/save-article-to-list-variant.js (hunk 1 above)
Comment (gyss) [reply myft/ui/save-article-to-list-variant.js:143]: In order to improve accessibility, this form control should have a hidden label with the text we want to read associated with the input. e.g.:

```
<label for="list-name" class="o-normalise-visually-hidden">Insert list name:</label>
<input type="text" name="list-name" />
```

File: myft/ui/save-article-to-list-variant.js (hunk 2 above)
Comment (jamesnicholls) [reply myft/ui/save-article-to-list-variant.js]: Thanks @gyss. I agree the order VoiceOver announces the interactive elements is weird. However, I'm not sure we can do much about it, as the structure is created by the Origami overlay component.

**Review by jamesnicholls — APPROVED** [review]
The live region works great. I think all the accessibility concerns have been addressed now

**Review by gyss — APPROVED** [review]
I understand the screenshots are not updated. I've checked the new version and there are a couple of minor accessibility issues. Anyways looks like those are related to the component imported as @jamesnicholls said.

**Comment by meel-io** [comment]
Thanks for your help and your feedback 💯

**Pull request merged**